src/openrct2-ui/windows/Banner.cpp

```diff
@@ -69,264 +69,218 @@ static rct_widget window_banner_widgets[] = {
-static void window_banner_viewport_rotate(rct_window *w);
-static void window_banner_invalidate(rct_window *w);
-static void window_banner_paint(rct_window *w, rct_drawpixelinfo *dpi);
-
-static rct_window_event_list window_banner_events([](auto& events)
-{
-    events.mouse_up = &window_banner_mouseup;
-    events.mouse_down = &window_banner_mousedown;
-    events.dropdown = &window_banner_dropdown;
-    events.text_input = &window_banner_textinput;
-    events.viewport_rotate = &window_banner_viewport_rotate;
-    events.invalidate = &window_banner_invalidate;
-    events.paint = &window_banner_paint;
-});
 // clang-format on
 
-/**
- *
- *  rct2: 0x006BA305
- */
-rct_window* window_banner_open(rct_windownumber number)
+class BannerWindow final : public Window
 {
-    rct_window* w;
-    rct_widget* viewportWidget;
+private:
+    Banner* banner;
+    CoordsXYZ bannerViewPos;
+    TileElement* tileElement = nullptr;
 
-    // Check if window is already open
-    w = window_bring_to_front_by_number(WC_BANNER, number);
-    if (w != nullptr)
-        return w;
-
-    w = WindowCreateAutoPos(WW, WH, &window_banner_events, WC_BANNER, WF_NO_SCROLLING);
-    w->widgets = window_banner_widgets;
-    w->enabled_widgets = (1 << WIDX_CLOSE) | (1 << WIDX_BANNER_TEXT) | (1 << WIDX_BANNER_NO_ENTRY) | (1 << WIDX_BANNER_DEMOLISH)
-        | (1 << WIDX_MAIN_COLOUR) | (1 << WIDX_TEXT_COLOUR_DROPDOWN) | (1 << WIDX_TEXT_COLOUR_DROPDOWN_BUTTON);
-
-    w->number = number;
-    WindowInitScrollWidgets(w);
-
-    auto banner = GetBanner(w->number);
-    auto bannerViewPos = banner->position.ToCoordsXY().ToTileCentre();
+    void CreateViewport()
+    {
+        rct_widget* viewportWidget = &window_banner_widgets[WIDX_VIEWPORT];
+        viewport_create(
+            this, windowPos + ScreenCoordsXY{ viewportWidget->left + 1, viewportWidget->top + 1 },
+            (viewportWidget->width()) - 1, (viewportWidget->height()) - 1, 0, bannerViewPos, 0, SPRITE_INDEX_NULL);
+
+        if (viewport != nullptr)
+            viewport->flags = gConfigGeneral.always_show_gridlines ? VIEWPORT_FLAG_GRIDLINES : 0;
+        Invalidate();
+    }
 
-    TileElement* tile_element = map_get_first_element_at(bannerViewPos);
-    if (tile_element == nullptr)
-        return nullptr;
-    while (1)
+    void InitTileElement()
     {
-        if ((tile_element->GetType() == TILE_ELEMENT_TYPE_BANNER) && (tile_element->AsBanner()->GetIndex() == w->number))
+        TileElement* tile_element = map_get_first_element_at(banner->position.ToCoordsXY().ToTileCentre());
+        if (tile_element == nullptr)
         {
-            break;
+            tileElement = nullptr;
+            return;
         }
 
-        tile_element++;
+        while (1)
+        {
+            if ((tile_element->GetType() == TILE_ELEMENT_TYPE_BANNER) && (tile_element->AsBanner()->GetIndex() == number))
+                break;
+            tile_element++;
+        }
+        tileElement = tile_element;
```

Comment:
The original code wasn't very good, this while loop is quite dangerous as it will traverse other tiles or even bad memory if the banner position is incorrect. Can you make it bail the loop if it reaches the end of the tile?